packages/flutter/lib/src/widgets/routes.dart

```diff
@@ -1602,11 +1602,57 @@ abstract class ModalRoute<T> extends TransitionRoute<T> with LocalHistoryRoute<T
   ///  * [Form], which provides an `onWillPop` callback that uses this mechanism.
   ///  * [addScopedWillPopCallback], which adds callback to the list
   ///    checked by [willPop].
+  @Deprecated(
+    'Use registerPopInterface or PopScope instead. '
+    'This feature was deprecated after v3.12.0-1.0.pre.',
+  )
   void removeScopedWillPopCallback(WillPopCallback callback) {
     assert(_scopeKey.currentState != null, 'Tried to remove a willPop callback from a route that is not currently in the tree.');
     _willPopCallbacks.remove(callback);
   }
 
+  /// Registers the existence of a [PopScope] widget in the route.
+  ///
+  /// [PopScope] widgets registered in this way will have their
+  /// [PopScope.onPopInvoked] callbacks called when a route is popped or a pop is
+  /// attempted. They will also be able to block pop operations with
+  /// [PopScope.popEnabled] through this route's [popEnabled] method.
+  ///
+  /// See also:
+  ///
+  ///  * [unregisterPopInterface], which performs the opposite operation.
+  void registerPopInterface(PopInterface popInterface) {
+    _popInterfaces.add(popInterface);
+    popInterface.popEnabledNotifier.addListener(_updateSystemNavigator);
+    _updateSystemNavigator();
+  }
+
+  /// Unregisters a [PopScope] widget in the route's widget subtree.
+  ///
+  /// See also:
+  ///
+  ///  * [registerPopInterface], which performs the opposite operation.
+  void unregisterPopInterface(PopInterface popInterface) {
+    _popInterfaces.remove(popInterface);
+    popInterface.popEnabledNotifier.removeListener(_updateSystemNavigator);
+    _updateSystemNavigator();
+  }
+
+  // Tells the SystemNavigator whether or not a system pop should have an
+  // effect.
+  void _updateSystemNavigator() {
+    if (!isCurrent) {
+      return;
+    }
+    final NavigationNotification notification = NavigationNotification(
+      // canPop indicates that the originator of the Notification can handle a
+      // pop. In the case of PopScope, it handles pops when popEnabled is
+      // false. Hence the seemingly backward logic here.
+      canHandlePop: popEnabled() == RoutePopDisposition.doNotPop,
```

Comment:
we should think about debouncing. This notification will be fired every time a popscope is register and unregister, if people start to use it a lot, you can see an influx of notifications. I am not sure if this is a serious issue or not.

I kind of have the same issue in LocalHistoryRoute._entriesImpliesAppBarDismissal, where i use a counter to fix the problem.

Reply:
One good thing is that, if I'm remembering correctly, a dispatched Notification only traverses a separate NotificationListener tree and not the full Widget tree.  I think this came up in the design doc.

Also, now that I'm using a ValueNotifier for PopScope.canPop, this won't be dispatched unless the value changes.  So in the case of a PopScope in a widget subtree that's animating and is being rebuilt 60 times per second, the notification won't get dispatched unless canPop changes.